buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/Reaper.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle.reaper;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A standalone process that will reap external services after a build dies.
+ *
+ * <h2>Input</h2>
+ * Since how to reap a given service is platform and service dependent, this tool
+ * operates on system commands to execute. It takes a single argument, a directory
+ * that will contain files with reaping commands. Each line in each file will be
+ * executed with {@link Runtime#getRuntime()#exec}.
+ *
+ * The main method will wait indefinitely on the parent process (Gradle) by
+ * reading from stdin. When Gradle shuts down, whether normally or abruptly, the
+ * pipe will be broken and read will return.
+ *
+ * The reaper will then iterate over the files in the configured directory,
+ * and execute the given commands. If any commands fail, a failure message is
+ * written to stderr. Otherwise, the input file will be deleted. If no inputs
+ * produced errors, the entire input directory is deleted upon completion of reaping.
+ */
+public class Reaper implements Closeable {
+
+    private Path inputDir;
+    private boolean failed;
+
+    private Reaper(Path inputDir) {
+        this.inputDir = inputDir;
+        this.failed = false;
+    }
+
+    public static void main(String[] args) throws Exception {
+        if (args.length != 1) {
+            System.err.println("Expected one argument.\nUsage: java -jar reaper.jar <DIR_OF_REAPING_COMMANDS>");
+            System.exit(1);
+        }
+        Path inputDir = Paths.get(args[0]);
+
+        try (Reaper reaper = new Reaper(inputDir)){
+            System.in.read();
+            reaper.reap();
+        }
+    }
+
+    private void reap() {
+        try {
+            final List<Path> inputFiles = Files.list(inputDir)
```

Comment:
I explicitly want to collect all these before we start hte loop. Using an active stream over the list will mean we can pick up changes to the files. This was more important to avoid in a first iteration when I dynamically added the log file, but I do like iterating over something that can't change so I will keep this as is.

Reply:
What I meant is that per the docs  `Files.list` returns a `Stream ` that  needs to be closed and we are not doing that here. You can still collect it to a list first, just make sure to close the stream.

Reply:
I don't think it is necessary. From the Stream docs:

```
Streams have a BaseStream.close() method and implement AutoCloseable, but nearly all stream instances do not actually need to be closed after use. Generally, only streams whose source is an IO channel (such as those returned by Files.lines(Path, Charset)) will require closing. Most streams are backed by collections, arrays, or generating functions, which require no special resource management.
```

Doing a filesystem operation here is not actually doing IO (it's only filesystem metadata).

Reply:
I was looking at [Files.list](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#list-java.nio.file.Path-) and [DirectoryStream](https://docs.oracle.com/javase/8/docs/api/java/nio/file/DirectoryStream.html) specifically ```Failure to close the stream may result in a resource leak.```

Reply:
TIL (!). Thanks @atorok. I pushed c057be6